### ISSUE TITLE
False negative, no warning when "x is not defined" diagnostic should be displayed

### ISSUE THREAD

**Issue by matthewcarbone** [microsoft/pyright#7564]
I'm honestly not even sure if this is a Pyright issue or not but I could use a bit of help here. Pyright has been working perfectly for me _except_ for one really annoying issue. The most minimal example I can come up with is as follows.

```python
def foo():
  x = x2  # [reportUndefinedVariable] displayed, correctly, as x2 is undefined
  return x
```

```python
def foo():
  x = x  # no diagnostic displayed, even though x is not defined in any scope
  return x
```

Any idea what's going on here? Apologies if this is something obvious and I've missed something simple, but googling around for 10 minutes has yielded nothing, and I'm not even sure what to search for. Thanks!

**Comment by erictraut** [comment]
I'm not able to repro the problem. Pyright generates a `reportUnboundVariable` diagnostic for me as expected. I suspect you've disabled this diagnostic check in your configuration. Please try with a fresh project and a default configuration.

**Comment by matthewcarbone** [comment]
@erictraut yup you're absolutely correct. Sorry for the bother. I'll need to fiddle around with the settings. Thanks!

**Comment by matthewcarbone** [comment]
@erictraut curiously, this appears to be due to setting `typeCheckingMode = "off"` in my config. Turning it back to "basic" "fixes" the problem (but introduces a host of other "errors" due to lack of stubs, I think). That said, any ideas why this discrepancy is being caused by this option? I'm not sure what this has to do with type checking.

**Comment by erictraut** [comment]
Refer to [this documentation](https://microsoft.github.io/pyright/#/configuration?id=diagnostic-settings-defaults) for a table of which diagnostic rules are enabled by default based on the `typeCheckingMode`.

**Comment by matthewcarbone** [comment]
That works, thank you. For anyone coming here after, the seemingly correct way to set this up is by using `diagnosticSeverityOverrides`. So, something like this (in lazy neovim):

```lua
lspconfig["pyright"].setup({
    capabilities = capabilities,
    on_attach = on_attach,
    settings = {
        pyright = {
            disableLanguageServices = false,
            disableOrganizeImports = false,
        },
        python = {
            analysis = {
                typeCheckingMode = "off",
                useLibraryCodeForTypes = false,
                autoImportCompletions = true,
                diagnosticSeverityOverrides = {
                    reportUnusedVariable = "warning",
                    reportUnboundVariable = "error",
                },
            },
        },
    },
})
```

**Comment by erictraut** [comment]
Yes, you can use `diagnosticSeverityOverrides` in the language server settings. Or you can create a pyrightconfig.json or pyproject.toml configuration file in your project. If your project is shared among multiple developers, a config file is preferable because it can be checked in to the repo, and everyone will see the same settings.
